### ISSUE TITLE
prefer-for-of rule false positive

### ISSUE THREAD

**Issue by plantain-00** [palantir/tslint#1931]
### Bug Report

- __TSLint version__: `4.2.0`
- __TypeScript version__: `2.1.4`
- __Running TSLint via__: (please choose one) CLI

#### TypeScript code being linted

```ts
const a = [1, 2];
const b = [2, 3];
for (let i = 0; i < a.length; i++) {
    console.log(a[i] + b[i]);
}
```

with `tslint.json` configuration:

```json
{
    "extends": "tslint:latest",
    "rules": {
        "forin": false,
        "interface-name": [
            false
        ],
        "max-line-length": [
            false
        ],
        "no-var-requires": false,
        "no-console": [
            false
        ],
        "no-string-literal": false,
        "no-reference": false,
        "ordered-imports": [
            false
        ],
        "object-literal-sort-keys": false,
        "variable-name": [
            true,
            "ban-keywords"
        ],
        "no-bitwise": false,
        "member-access": false,
        "arrow-parens": false,
        "array-type": [true, "array"],
        "max-classes-per-file": [false]
    }
}
```

#### Actual behavior

Expected a 'for-of' loop instead of a 'for' loop with this simple iteration

#### Expected behavior

No error.
Because the code cannot be transformed to for...of.

**Comment by adidahiya** [comment]
if you have support for `Array.prototype.entries()` in your environment, this could work for you:

```ts
for (const [i, elem] of a.entries()) {
    console.log(elem + b[i]);
}
```

but you're right, this looks like a bug in the rule implementation; it probably shouldn't flag that code as a failure.

**Comment by nchen63** [comment]
I can't reproduce this issue. Can anyone else?

**Comment by plantain-00** [comment]
@nchen63 
Another example closer to my real code:
```ts
declare const value1: any;
declare const value2: any;
for (let i = 0; i < (value1 as number[]).length; i++) {
    if ((value1 as number[])[i] === (value2 as number[])[i]) {
    }
}
```

**Comment by nchen63** [comment]
@plantain-00 That one reproduces the issue

**Comment by plantain-00** [comment]
Sorry for initial wrong code, I have no clue how it happened then.
Thanks for fixing this.
